### DISCUSSION TITLE
Multiple row level subscriptions in one channel?

### DISCUSSION THREAD

**Discussion by lubeskih** [supabase/supabase#10980] [Questions] [answered]
Hi. I am using Supabase Realtime feature on the frontend with TypeScript. I have a client that creates one channel on init:

```
this.channel = this.supabaseClient.channel("some_channel");
```

And afterwards I subscribe to UPDATE on different rows in the table of choice:

```
  public async subscribe(uniqueValue: string) {
    const subscription = this.channel
      .on(
        "postgres_changes",
        {
          event: "UPDATE",
          schema: "public",
          table: "my_table",
          filter: `some_column=eq.${uniqueValue}`,
        },
        (payload) => {...}
      )
      .subscribe();

   // keep track of active subscriptions to be able to unsubscribe later
     this.activeSubs.set(uniqueValue, subscription);
    return;
  }
  ``` 

Additionally, when I want to unsubscribe I do:

```
  public async unsubscribe(uniqueValue: string) {
    if (this.activeSubs.has(uniqueValue)) {
      await this.activeSubs.get(uniqueValue)?.unsubscribe();
      this.activeSubs.delete(uniqueValue);

      return;
    }
  }
```

And this works fine for one subscription. But if I call my `subscribe` function one after another (subscribe on same channel but two different rows), I get

```
Uncaught (in promise) tried to subscribe multiple times. 'subscribe' can only be called a single time per channel instance
```

My question is: How do I make multiple row level subscriptions on one channel? Is this possible, or I must make a channel for each row level subscription? Additionally, if one user makes 100 subscriptions, that's 100 channels. Multiplied by number of users of the app, this can go to thousands of channels. Will my application's performance be affected? 

Thank you. :-)

**Comment by lubeskih** [comment, marked as answer]
I found a solution to my question. I guess all "issues" boil down to the fact that I don't read the documentation carefully.

> A [channel](https://hexdocs.pm/phoenix/channels.html) is the basic building block of Realtime and narrows the scope of data flow to subscribed clients. You can think of a channel as a chatroom where participants are able to see who's online and send and receive messages; similar to a Discord or Slack channel.

So to subscribe on multiple row level changes, you would have a channel for each row:

```
supabase
  .channel('public:countries:id=eq.200')
  .on('postgres_changes', { event: 'UPDATE', schema: 'public', table: 'countries', filter: 'id=eq.200' }, handleRecordUpdated)
  .subscribe()
```

So in my case, I tweaked my `subscribe` function like so:

```
...
   const subscription = this.supabaseClient
      .channel(`public:table:some_column=eq.${uniqueValue}`)
     .on(...)
...
```

So, the way I understand this: 1000 rows = 1000 open channels if at least one client listens on each rows. If more clients start listening on a row, they will just use the existing channel created. Is this correct?

**Reply by w3b6x9** [reply]
@lubeskih channel name must be unique but beyond that you can also chain so that a single channel can listen to multiple filters:

```
...
   const subscription = this.supabaseClient
      .channel(`channel-a`)
     .on(...)
     .on(...)
     .on(...)
...
```

**Reply by lubeskih** [reply]
@w3b6x9 epic, thanks. :-)

**Reply by Donnerstagnacht** [reply]
Is there a solution for a case if I would like subscribe once the user visits a page, unsubscribe if the users leaves the page and subscribes again if the user visits a second time?

```
That causes the following error in my application: 
ERROR tried to subscribe multiple times. 'subscribe' can only be called a single time per channel instance

```
